### PR TITLE
Add observers for stylesheet mutations

### PR DESCRIPTION
**Pull request by dcramer**
This captures the insertRule and deleteRule actions on CSSStyleSheet, and wires them up to StyleSheetRule events in the recorder.

### CODE REVIEW

**Comment by dcramer** [comment]
We should squash this for sure, but let me know if the approach is how you'd go about it.

**Comment by dcramer** [comment]
Also this fixes GH-176

**Comment by dcramer** [comment]
cc @marcospassos

File: test/integration.test.ts [hunk 1]
```diff
@@ -35,14 +35,7 @@ describe('record integration tests', function(this: ISuite) {
...
   before(async () => {
-    this.browser = await puppeteer.launch({
-      defaultViewport: {
-        width: 1920,
-        height: 1080,
-      },
-      headless: false,
-      args: ['--no-sandbox'],
-    });
+    this.browser = await launchPuppeteer();
```
Comment (dcramer) [inline test/integration.test.ts]: these changes are unrelated, but I was annoyed with running --watch and having the browser constantly take focus

File: test/__snapshots__/record.test.ts.snap [hunk 2]
```diff
@@ -331,3 +331,130 @@ exports[`custom-event 1`] = `
...
+        \\"left\\": 0,
+        \\"top\\": 0
+      }
+    }
+  },
+  {
+    \\"type\\": 3,
+    \\"data\\": {
+      \\"source\\": 8,
+      \\"id\\": -1,
```
Comment (dcramer) [inline test/__snapshots__/record.test.ts.snap]: guessing the use of the mapper here is wrong, could someone guide me on the expectation of id and how it needs to function?

**Comment by dcramer** [comment]
Also looks like the test suite is a bit flakey (sometimes the async code doesnt finish running as expected?)

Have seen it on the new tests + on:

```
     AssertionError: Received value does not match stored snapshot "async-checkout 1".
```

File: package.json [hunk 3]
```diff
@@ -4,6 +4,7 @@
   "description": "record and replay the web",
   "scripts": {
     "test": "npm run bundle:browser && cross-env TS_NODE_CACHE=false TS_NODE_FILES=true mocha -r ts-node/register test/**/*.test.ts",
+    "test:watch": "PUPPETEER_HEADLESS=true npm run test -- --watch --watch-extensions js,ts",
```
Comment (dcramer) [inline package.json]: can back this out if folks disagree but found it useful

aside, it doesnt recompile src/ and I couldn't figure out how to achieve that

File: src/record/observer.ts [hunk 4]
```diff
@@ -519,6 +520,34 @@ function initInputObserver(
   };
 }
 
+function initStyleSheetObserver(cb: styleSheetRuleCallback): listenerHandler {
+  const insertRule = CSSStyleSheet.prototype.insertRule;
+  CSSStyleSheet.prototype.insertRule = function(
+    rule: string,
+    index?: number | undefined,
+  ) {
+    cb({
```
Comment (dcramer) [inline src/record/observer.ts]: one question I had here - do people prefer this version of a single type, or should we have two types which means we could better compress the data

Comment (Yuyz0112) [reply src/record/observer.ts]: I think this implementation keeps simplicity. Data compressing stuff can be done by the packer plugin in a more general aspect.

File: src/record/observer.ts [hunk 5]
```diff
@@ -519,6 +520,34 @@ function initInputObserver(
   };
 }
 
+function initStyleSheetObserver(cb: styleSheetRuleCallback): listenerHandler {
+  const insertRule = CSSStyleSheet.prototype.insertRule;
+  CSSStyleSheet.prototype.insertRule = function(
+    rule: string,
+    index?: number | undefined,
```
Comment (Yuyz0112) [inline src/record/observer.ts]: looks like `index?: number` is enough since `?` already make type optional

File: package.json (hunk 3 above)
Comment (Yuyz0112) [reply package.json]: Could not get that done too...
Combine Mocha and Typescript is not as fluent as jest.

**Comment by Yuyz0112** [comment]
@dcramer Thanks a lot! I would love to merge it after tweak some small changes.

The test things can be done in the following PRs.

**Comment by dcramer** [comment]
@yuzy0112 I fixed the type - were you asking me to pull the test:watch and other changes out?

**Comment by Yuyz0112** [comment]
Nope, it's fine.

**Pull request merged**